Jetpack Focus: Display phase one overlays for Stats, Reader, and Notifications

hassaanelgarem opened the pull request:
Part of #19447

## Description
This PR displays the feature-specific overlay for phase one for Stats, Reader, and Notification.

## Notes

- Track events will be handled in a future PR.
- The action for the Switch button is still not implemented. It will be implemented under another project.

### Screenshots

| / | Stats | Reader | Notifications |
| - | - | - | - |
| Light Mode |![stats-light](https://user-images.githubusercontent.com/25306722/200698331-62beb147-093b-43df-b043-95bb917a1854.png)|![reader-light](https://user-images.githubusercontent.com/25306722/200698343-d6fbd301-7128-4f72-8f69-9b608b341a10.png)|![notifs-light](https://user-images.githubusercontent.com/25306722/200698345-bc02de1c-b019-4a90-87af-392b54e55530.png)|
| Dark Mode |![stats-dark](https://user-images.githubusercontent.com/25306722/200698392-28fa8d72-adf8-4ec4-9d8c-84a4d7ff06df.png)|![reader-dark](https://user-images.githubusercontent.com/25306722/200698396-07266c68-5a4a-40ea-860e-274c18ff5efa.png)|![notifs-dark](https://user-images.githubusercontent.com/25306722/200698401-df6855e1-8ba0-46dd-94b6-3e861c56ed22.png)|

## Testing Instructions
### Phase one disabled
1. Open the app
2. Make sure the "Jetpack Features Removal Phase one" flag is disabled from the debug menu
3. Navigate to Stats
4. Make sure that the overlay is not displayed
5. Navigate to Notifications 
6. Make sure that the overlay is not displayed
7. Navigate to the Reader
8. Make sure that the overlay is not displayed

### Phase one enabled
P.S: To be able to test, you can disable the frequency logic by changing the line in `JetpackFeaturesRemovalCoordinator.swift:20` to be `return .defaultConfig`

1. Open the app
2. Make sure the "Jetpack Features Removal Phase one" flag is enabled from the debug menu
3. Navigate to stats
4. Make sure the overlay is displayed with the correct content and matches the design
5. Turn on dark mode
6. Make sure the overlay matches the dark mode design
7. Repeat the above steps for Reader and Notifications

### Overlay Frequencies
P.S: The general and feature-specific frequencies for phase one are 2 and 7 days, respectively. To be able to test the frequency flows, you can do the following:
Change the value of `secondsInDay` in `JetpackOverlayFrequencyTracker.swift:86` from `86_000` to `60`, for example. This changes the frequencies to be 2 and 7 **minutes**. 

1. Fresh install the app (Delete then install)
2. Open the app
3. Navigate to stats
4. Make sure the overlay is presented
5. Before the general frequency has passed, navigate to Notifications
6. Make sure the overlay is not presented
7. After the general frequency has passed, navigate to Notifications
8. Make sure the overlay is presented
9. Wait till the general frequency has passed
10. Before the feature-specific frequency has passed, navigate to Notifications
11. Make sure the overlay is not presented
12. After the feature-specific frequency has passed, navigate to Notifications
13. Make sure the overlay is presented

## Regression Notes
1. Potential unintended areas of impact
N/A

2. What I did to test those areas of impact (or what existing automated tests I relied on)
N/A

5. What automated tests I added (or what prevented me from doing so)
N/A

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

hassaanelgarem committed:
Update: abstract out jetpack overlay config to be a protocol

hassaanelgarem committed:
Add: configure should show for general phase

hassaanelgarem committed:
Add: configure title for phase 1

hassaanelgarem committed:
Add: configure subtitle for phase one

hassaanelgarem committed:
Add: feature specific Lottie animation files

hassaanelgarem committed:
Remove: Lottie files from the Jetpack bundle

hassaanelgarem committed:
Add: configure animations for phase one

hassaanelgarem committed:
Add: configure footnote for phase one

hassaanelgarem committed:
Add: configure analytics source for all phases

hassaanelgarem committed:
Add: configure learn more button for phase one

hassaanelgarem committed:
Add: configure switch button title for phase one

hassaanelgarem committed:
Add: configure continue button title for phase one

hassaanelgarem committed:
Add: configure close button for phase one

hassaanelgarem committed:
Refactor: animation file names to constants

hassaanelgarem committed:
Add: Func to present overlay in `JetpackFeaturesRemovalCoordinator`

hassaanelgarem committed:
Add: overlay triggers for stats, reader and notifs

hassaanelgarem committed:
Remove: overlay site creation config for now

hassaanelgarem committed:
Add: overlay frequency logic

hassaanelgarem committed:
Fix: Error in frequency tracker's logic

hassaanelgarem committed:
Test: remove the overlay frequency logic

shaunandrews commented:
I wasn't able to get this running on my iOS device, but based on the screenshots I think there's some sizing and spacing issues. Here's a side-by-side, your screenshot on the left and the Figma design on the right:

<img width="1573" alt="CleanShot 2022-11-09 at 12 06 19@2x" src="https://user-images.githubusercontent.com/191598/200894890-ec01e693-63e3-4b13-9ef7-40ed6a9bebbf.png">

The logos look too big, and the line-height for the heading seems to large.

hassaanelgarem committed:
Fix: faulty frequency logic

hassaanelgarem committed:
Merge branch 'task/19447-phase-1' into task/19447-phase-1-test

hassaanelgarem committed:
Update: Animation view dimensions to match the design

hassaanelgarem committed:
Update: title label to match the design

hassaanelgarem committed:
Merge branch 'task/19447-phase-1' into task/19447-phase-1-test

hassaanelgarem commented:
@shaunandrews The logos were respecting the size defined in the Lottie animation. I've changed that to have a static height instead. I've also updated the title's line-height to match the design.

<img src="https://user-images.githubusercontent.com/25306722/200988456-838b234e-a1d2-4284-abeb-3a4a5d075235.png" width=40%>


Let me know if you notice anything else!

hassaanelgarem committed:
Update: Overlay spacings in portrait

hassaanelgarem committed:
Revert "Test: remove the overlay frequency logic"

This reverts commit 03ce278da80bbb5013728c0ac72d7f82303fb1f6.

hassaanelgarem committed:
Update: disable dismissing the overlay by swiping down

hassaanelgarem committed:
Merge branch 'trunk' into task/19447-phase-1

momo-ozawa commented on WordPress/Classes/ViewRelated/Jetpack/Branding/Overlay/JetpackFullscreenOverlayGeneralPhaseConfig.swift:
```diff
@@ -0,0 +1,221 @@
+                                                     value: "Follow any site with the Jetpack app",
+                                                     comment: "Title of a screen displayed when the user accesses the Reader screen from the WordPress app. The screen showcases the Jetpack app.")
+                static let subtitle = NSLocalizedString("jetpack.fullscreen.overlay.phaseOne.reader.subtitle",
+                                                     value: "Switch to the Jetpack app to find, fillow, and like all your favorite sites and posts with Reader.",
```
[nit] typo
```suggestion
                                                     value: "Switch to the Jetpack app to find, follow, and like all your favorite sites and posts with Reader.",
```

hassaanelgarem replied:
Thanks for catching this! Fixed in e3139826b55785ab1cca29cfda1330a2b58df399 👍

momo-ozawa commented on WordPress/Classes/ViewRelated/Jetpack/Branding/Coordinator/JetpackFeaturesRemovalCoordinator.swift:
```diff
@@ -61,4 +83,27 @@ class JetpackFeaturesRemovalCoordinator {
+    /// - Parameters:
+    ///   - source: The source that triggers the display of the overlay.
+    ///   - viewController: View controller where the overlay should be presented in.
+    static func presentOverlay(from source: OverlaySource, in viewController: UIViewController) {
```
I wonder if it would make more sense to call this `presentOverlayIfNeeded(from: in:)`. I'll leave the decision up to you though :)

hassaanelgarem replied:
Agreed! Updated in 6426d1a9e2e273703cfebadfb152711795168a59

momo-ozawa reviewed (approved):
Works as described w 2 minor comments

hassaanelgarem committed:
Fix: typo in reader's overlay subtitle

hassaanelgarem committed:
Refactor: rename `presentOverlay` to `presentOverlayIfNeeded`

shaunandrews commented:
I gave this another spin; It feels pretty good to me.

--

A very minor thing I did notice (perhaps for another PR) is the color of the "X" button in dark mode; It feels too bright when [compared to the design in Figma](https://www.figma.com/file/qk5s4q4mZsXknwhMWo2vIY/WordPress--%3E-Jetpack-Migration?node-id=5194%3A17264&t=fTYJqHBT4wsvhzG7-4):

<img width="447" alt="CleanShot 2022-11-15 at 13 35 38@2x" src="https://user-images.githubusercontent.com/191598/201999339-bc823721-28ff-48e8-a39b-d7b2c0423675.png">

I'm not sure if this is related to just this overlay or PR — its possible that this is an issue with other uses of the overlay.

hassaanelgarem commented:
> I'm not sure if this is related to just this overlay or PR — its possible that this is an issue with other uses of the overlay.

@shaunandrews You're right. I've used the same settings for the bottom sheet overlay.

I created [this issue](https://github.com/wordpress-mobile/WordPress-iOS/issues/19599) to track this, and I'll update this for all overlays in a separate future PR 👍
